client/alloc_watcher.go

```diff
@@ -0,0 +1,448 @@
+package client
+
+import (
+	"archive/tar"
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/hashicorp/consul/lib"
+	nomadapi "github.com/hashicorp/nomad/api"
+	"github.com/hashicorp/nomad/client/allocdir"
+	"github.com/hashicorp/nomad/client/config"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// rpcer is the interface needed by a prevAllocWatcher to make RPC calls.
+type rpcer interface {
+	// RPC allows retrieving remote allocs.
+	RPC(method string, args interface{}, reply interface{}) error
+}
+
+// terminated is the interface needed by a prevAllocWatcher to check if an
+// alloc is terminated.
+type terminated interface {
+	Terminated() bool
+}
+
+// prevAllocWatcher allows AllocRunners to wait for a previous allocation to
+// terminate and migrate its data whether or not the previous allocation is
+// local or remote.
+type prevAllocWatcher interface {
+	// Wait for previous alloc to terminate
+	Wait(context.Context) error
+
```

Comment:
Add `IsWaiting` and `IsMigrating` methods and the alloc runner can just call into the watcher so we can skip storing that state there and remove the locks